src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/CodeBasedDependencyAlgorithm.cs

```diff
@@ -118,23 +120,82 @@ public static void AddDependenciesDueToMethodCodePresence(ref DependencyList dep
-                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetClosedDelegateMarshallingStub(parameter)), "Delegate Marshalling Stub");
-                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetForwardDelegateCreationStub(parameter)), "Delegate Marshalling Stub");
+            if (metadataType != null)
+            {
+                if (metadataType.Name == "Marshal" && metadataType.Namespace == "System.Runtime.InteropServices")
+                {
+                    foreach (TypeDesc type in method.Instantiation)
+                    {
+                        AddPInvokeParameterDependencies(ref dependencies, factory, type);
+                    }
+                }
             }
-            else if (Internal.TypeSystem.Interop.MarshalHelpers.IsStructMarshallingRequired(parameter))
+        }
+
+        public static void AddPInvokeParameterDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc parameter)
+        {
+            AddDependenciesDuePInvokeDelegate(ref dependencies, factory, parameter);
+            AddDependenciesDuePInvokeStruct(ref dependencies, factory, parameter);
+        }
+
+        public static void AddDependenciesDuePInvokeDelegate(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
+        {
+            if (type.IsDelegate)
+            {
+                dependencies.Add(factory.NecessaryTypeSymbol(type), "Delegate Marshalling Stub");
+
+                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetOpenStaticDelegateMarshallingStub(type)), "Delegate Marshalling Stub");
+                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetClosedDelegateMarshallingStub(type)), "Delegate Marshalling Stub");
+                dependencies.Add(factory.MethodEntrypoint(factory.InteropStubManager.GetForwardDelegateCreationStub(type)), "Delegate Marshalling Stub");
+            }
+        }
+
+        public static void AddDependenciesDuePInvokeStruct(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
+        {
+            if (MarshalHelpers.IsStructMarshallingRequired(type))
             {
-                var stub = (Internal.IL.Stubs.StructMarshallingThunk)factory.InteropStubManager.GetStructMarshallingManagedToNativeStub(parameter);
-                dependencies.Add(factory.ConstructedTypeSymbol(factory.InteropStubManager.GetStructMarshallingType(parameter)), "Struct Marshalling Type");
+                dependencies.Add(factory.NecessaryTypeSymbol(type), "Struct Marshalling Stub");
+
+                var stub = (Internal.IL.Stubs.StructMarshallingThunk)factory.InteropStubManager.GetStructMarshallingManagedToNativeStub(type);
+                dependencies.Add(factory.ConstructedTypeSymbol(factory.InteropStubManager.GetStructMarshallingType(type)), "Struct Marshalling Type");
```

Comment:
> For example, here 

What will be the position of `McgModuleManager` in the CoreRT world? If getting the native struct size is the only thing we need the EEType for, the code in question could be replaced by a lookup in e.g. the `struct_marshalling_stub_map` hashtable - instead of placing a pointer to the necessary type of the native struct type, you could just directly emit the native struct size into the hashtable. It will save way more than 32 bytes per marshallable struct we would otherwise waste on the EEType and the indirection to it.